File: Leetcode/Python/_20.py
class Solution:
    def isValid(self, s: str) -> bool:
        temp = []
        hashMap = {"(": ")", "[": "]", "{": "}"}
        for c in s:
            if c in hashMap:
                temp.append(hashMap[c])
            elif len(temp) > 0 and c == temp[-1]:
                temp.pop()
            else:
                return False
        return len(temp) == 0

File: Leetcode/cpp/20.cpp
class Solution {
public:
    bool isValid(string s) {
        std::map<char, char> map = {{'(', ')'}, {'{', '}'}, {'[', ']'}};
        std::vector<char> query;
        for (auto c: s) {
          if (map.find(c) != map.end()) {
            query.push_back(map[c]);
          } else if (!query.empty() && query.back() == c) {
            query.pop_back();
          } else {
            return false;
          }
        }
      return query.empty() ? true : false;
    }
};

class SolutionV2 {
public:
    bool isValid(string s) {
        std::vector<char> stack;
        for (const auto& c : s) {
            if (c == '(') {
                stack.push_back(')');
            } else if (c == '[') {
                stack.push_back(']');
            } else if (c == '{') {
                stack.push_back('}');
            } else {
                if (!stack.empty() && stack.back() == c) {
                    stack.pop_back();
                    continue;
                } else {
                    return false;
                }
            }
        }
        return stack.empty();
    }
};
